### PR TITLE
Use XRPAmount instead of STAmount where possible

### PR DESCRIPTION
**Pull request by nbougalis**
Reviews: @ximinez, @vinniefalco, @scottschurr

### CODE REVIEW

File: src/ripple/app/ledger/Ledger.cpp [hunk 1]
```diff
@@ -452,7 +452,7 @@ bool Ledger::saveValidatedLedger (bool current)
         "INSERT OR REPLACE INTO Ledgers "
         "(LedgerHash,LedgerSeq,PrevHash,TotalCoins,ClosingTime,PrevClosingTime,"
         "CloseTimeRes,CloseFlags,AccountSetHash,TransSetHash) VALUES "
-        "('%s','%u','%s','%s','%u','%u','%d','%u','%s','%s');");
+        "('%s','%u','%s','%u','%u','%u','%d','%u','%s','%s');");
```
Comment (vinniefalco) [inline src/ripple/app/ledger/Ledger.cpp]: Was this intentional?

File: src/ripple/app/tx/impl/ApplyContext.h [hunk 2]
```diff
@@ -23,6 +23,7 @@
 #include <ripple/ledger/ApplyViewImpl.h>
 #include <ripple/core/Config.h>
 #include <ripple/protocol/STTx.h>
+#include <ripple/protocol/XRPAmount.h>
```
Comment (vinniefalco) [inline src/ripple/app/tx/impl/ApplyContext.h]: I think we might want to just add this to `<ripple/protocol/st.h>` and use that include instead of individually picking out items.

File: src/ripple/app/ledger/Ledger.cpp (hunk 1 above)
Comment (nbougalis) [inline src/ripple/app/ledger/Ledger.cpp]: I thought it was, although now it occurs to me that it may be wrong: `%u` is probably not the right type for `std::uint64_t`. I'll roll this bit back, nice catch.

File: src/ripple/app/tx/impl/SusPay.cpp [hunk 3]
```diff
@@ -183,23 +183,22 @@ SusPayCreate::doApply()
...
     auto const sle = ctx_.view().peek(
         keylet::account(account));
 
-    if (XRPAmount(STAmount((*sle)[sfBalance]).mantissa()) <
-        XRPAmount(ctx_.view().fees().accountReserve(
-            (*sle)[sfOwnerCount] + 1)))
-        return tecINSUFFICIENT_RESERVE;
+    // Check reserve and funds availability
+    {
+        auto const balance = STAmount((*sle)[sfBalance]).xrp();
```
Comment (vinniefalco) [inline src/ripple/app/tx/impl/SusPay.cpp]: Why not `auto const balance = (*sle)[sfBalance])` ?

File: src/ripple/app/tx/impl/SusPay.cpp [hunk 4]
```diff
@@ -183,23 +183,22 @@ SusPayCreate::doApply()
...
 
-    if (XRPAmount(STAmount((*sle)[sfBalance]).mantissa()) <
-        amount + XRPAmount(ctx_.view().fees().accountReserve(
-            (*sle)[sfOwnerCount] + 1)))
-        return tecUNFUNDED;
+        if (balance < reserve)
+            return tecINSUFFICIENT_RESERVE;
+
+        if (balance < reserve + STAmount(ctx_.tx[sfAmount]).xrp())
+            return tecUNFUNDED;
```
Comment (vinniefalco) [inline src/ripple/app/tx/impl/SusPay.cpp]: This is weird, why are we going through the `STAmount` conversion only to convert back to `XRPAmount`?

File: src/ripple/app/tx/impl/Transactor.cpp [hunk 5]
```diff
@@ -229,12 +229,12 @@ TER Transactor::apply ()
     }
 
     auto const& fees = view().fees();
-    mFeeDue = STAmount (getApp().getFeeTrack().scaleFeeLoad(
-        calculateBaseFee(), fees.base, fees.units, view().flags() & tapADMIN));
+    mFeeDue = getApp().getFeeTrack().scaleFeeLoad(
+        calculateBaseFee(), fees.base, fees.units, view().flags() & tapADMIN);
```
Comment (vinniefalco) [inline src/ripple/app/tx/impl/Transactor.cpp]: Isn't this supposed to be done elsewhere? Before it even gets into the open ledger?

File: src/ripple/app/tx/impl/Transactor.cpp [hunk 6]
```diff
@@ -229,12 +229,12 @@ TER Transactor::apply ()
...
 
     if (sle)
     {
-        mPriorBalance   = sle->getFieldAmount (sfBalance);
+        mPriorBalance   = STAmount ((*sle)[sfBalance]).xrp ();
```
Comment (vinniefalco) [inline src/ripple/app/tx/impl/Transactor.cpp]: I don't understand why we keep doing this conversion through `STAmount`

File: src/ripple/protocol/Rate.h [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+/** Represents a transfer rate
+
+    Transfer rates are specified as fractions of 1 billion. For example, a
+    transfer rate of 1% is represented as 1010000000.
+*/
+struct Rate
+    : private boost::totally_ordered <Rate>
+{
+    std::uint32_t rate;
```
Comment (vinniefalco) [inline src/ripple/protocol/Rate.h]: I think I would call this `value` to prevent expressions like `rate.rate`

File: src/ripple/app/tx/impl/SusPay.cpp (hunk 4 above)
Comment (nbougalis) [inline src/ripple/app/tx/impl/SusPay.cpp]: I didn't want `XRPAmount` to have a constructor _from_ `STAmount`. The code returns to us the field as an `STAmount`, so all we can do is call `.xrp()` on it.

Comment (vinniefalco) [inline src/ripple/app/tx/impl/SusPay.cpp]: Why can't you write `ctx_.tx[sfAmount].xrp()`?

File: src/ripple/app/tx/impl/SusPay.cpp (hunk 3 above)
Comment (nbougalis) [inline src/ripple/app/tx/impl/SusPay.cpp]: We could, but I wanted to leverage `XRPAmount` wherever possible and implicit conversions to `STAmount`

File: src/ripple/protocol/STAmount.h [hunk 8]
```diff
@@ -268,6 +280,10 @@ class STAmount
     {
         return (mValue == 0) && mIsNative;
     }
+
+    // Conversion to new-style amounts
```
Comment (vinniefalco) [inline src/ripple/protocol/STAmount.h]: pointless comment

File: src/ripple/app/tx/impl/SusPay.cpp (hunk 3 above)
Comment (vinniefalco) [inline src/ripple/app/tx/impl/SusPay.cpp]: Then why not `auto const balance = (*sle)[sfBalance].xrp()`?

File: src/ripple/protocol/STTx.h [hunk 9]
```diff
@@ -88,14 +88,6 @@ class STTx final
...
         return tx_type_;
     }
-    STAmount getTransactionFee () const
-    {
-        return getFieldAmount (sfFee);
-    }
-    void setTransactionFee (const STAmount & fee)
-    {
-        setFieldAmount (sfFee, fee);
-    }
```
Comment (vinniefalco) [inline src/ripple/protocol/STTx.h]: good riddance

File: src/ripple/protocol/XRPAmount.h [hunk 10]
```diff
@@ -25,6 +25,8 @@
 #include <beast/utility/Zero.h>
 #include <boost/operators.hpp>
 #include <cstdint>
+#include <string>
+#include <type_traits>
```
Comment (vinniefalco) [inline src/ripple/protocol/XRPAmount.h]: `<beast/cxx14/type_traits.h>`

File: src/ripple/protocol/XRPAmount.h [hunk 11]
```diff
@@ -55,15 +56,21 @@ class XRPAmount
         return *this;
     }
 
-    XRPAmount (std::int64_t value)
-        : value_ (value)
+    template <class Integer,
+        class = typename std::enable_if <
```
Comment (vinniefalco) [inline src/ripple/protocol/XRPAmount.h]: ...`class = std::enable_if_t<std::is_integral<Integer>::value>>`

Comment (nbougalis) [inline src/ripple/protocol/XRPAmount.h]: Nice - will change.

**Comment by ximinez** [comment]
Non-unity build breaks. `#include <algorithm>` and `#include <beast\cxx14\iterator>`?
```
IOUAmount.cpp
src\ripple\protocol\impl\IOUAmount.cpp(203) : error C2039: 'find_if' : is not a
[...]
member of 'std'
src\ripple\protocol\impl\IOUAmount.cpp(217) : error C2039: 'make_reverse_iterato
r' : is not a member of 'std'
[...]
```

**Comment by scottschurr** [comment]
Most of my comments are optional.  The only really important comment is that the new constructors and conversion methods really ought to have unit tests.  With those (once it builds non-unity) I'd give it a thumbs up.

**Comment by ximinez** [comment]
Well that was fun. I look forward to your followup changes.

**Comment by ximinez** [comment]
:+1:  The PR looks good. My concerns have been addressed, especially:
* Non-unity build works.
* IOU strings unit test - much than the one I offered.
* Checking properly for native fees.

File: src/ripple/rpc/impl/TransactionSign.cpp [hunk 12]
```diff
@@ -1083,12 +1083,21 @@ Json::Value transactionSubmitMultiSigned (
...
+        auto const fee = stpTrans->getFieldAmount (sfFee);
+
+        if (!isLegalNet (fee))
+        {
+            std::ostringstream err;
+            err << "Invalid " << sfFee.fieldName
+                << " field.  Fees must be specified in XRP.";
+            return RPC::make_error (rpcINVALID_PARAMS, err.str ());
+        }
+        if (fee <= 0)
```
Comment (scottschurr) [inline src/ripple/rpc/impl/TransactionSign.cpp]: As you noted earlier, this is a bug I introduced.  The `fee` may be zero in one special case of `RegularKeySet`.  Do you want to change this to `fee < 0`?  Or shall I do a separate pull request (since it's my bug).

Comment (nbougalis) [inline src/ripple/rpc/impl/TransactionSign.cpp]: No, I just forgot to fix this. I'll take care of it.

Comment (scottschurr) [inline src/ripple/rpc/impl/TransactionSign.cpp]: Gaaah.  Never mind.  Since it's multisigned I think there's always a fee associated with the signatures.  But I'm not positive.  I'll have to set up a test case and try it out.

**Comment by scottschurr** [comment]
It would be nice if there were unit tests for `STAmount::xrp()` and `STAmount::iou()`.  But they're not critical.  Nice change.
:+1:

File: src/ripple/rpc/impl/TransactionSign.cpp (hunk 12 above)
Comment (ximinez) [inline src/ripple/rpc/impl/TransactionSign.cpp]: I remember this discussion, and I confirmed that the zero-fee regular key has to be signed by the master key only, so this is correct unless we intend to change that behavior.

**Comment by nbougalis** [comment]
Requested unit tests added, rebased against develop. I'm marking this as *passed*.
